Tweak client-gen to support metrics API

DirectXMan12 opened the pull request:
This PR adds support to client-gen for readonly APIs as well as for customizing the resource name used for a given type.

This is required to support the clients generated for `k8s.io/metrics/pkg/apis/metrics`.

cc @caesarxuchao 

Currently based on #45180

**Release note**:
```release-note
NONE
```

DirectXMan12 committed:
[client-gen] Allow overriding the resource name

This commit allows client-gen to generate clients where the resource
name does not directly correspond to the type name.

The tag `// +resourceName=<name>` can be used to override the resource
name.

DirectXMan12 committed:
[client-gen] Support Read-Only APIs

This commit adds support for read-only APIs to client-gen.
If the tag `// +readonly=true` is used, only the {Get,List,Watch}
client methods will be generated.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45184)
<!-- Reviewable:end -->

DirectXMan12 commented:
@caesarxuchao this should be the requisite changes to get client generation working for k8s.io/metrics.

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/generator_for_type.go:
```diff
@@ -128,18 +134,22 @@ func (g *genClientForType) GenerateType(c *generator.Context, t *types.Type, w i
-		// Generate the UpdateStatus method if the type has a status
-		if genStatus(t) {
-			sw.Do(updateStatusTemplate, m)
+		if !readonly {
```
Could we reduce the level of nested indention, something like

```go
if !noMethods && !readonly{
 xxx
}

if !nomethods && readoly {
 xxx
}
```

caesarxuchao commented:
One nit.

Also you'll need to add a line in hack/update-codegen.sh if you want to keep metrics clientset up-to-date.

DirectXMan12 committed:
Enable client-gen for k8s.io/metrics

This commit turns on client-gen for k8s.io/metrics.  Clients are
generated for `k8s.io/metrics/pkg/apis/metrics` (both internal and
v1alpha1).  `k8s.io/metrics/pkg/apis/custom_metrics` uses a bespoke
client due to the unique nature of its paths.

DirectXMan12 commented:
nit should be addressed, and hack/update-codegen.sh has been updated.

caesarxuchao commented on staging/src/k8s.io/metrics/hack/update-codegen.sh:
```diff
@@ -0,0 +1,47 @@
+INPUT="--input ${INPUT_APIS[@]}"
+CLIENTSET_PATH="--clientset-path k8s.io/metrics/pkg/client/clientset_generated"
+
+${CLIENTGEN} ${INPUT_BASE} ${INPUT} ${CLIENTSET_PATH} --output-base ${KUBE_ROOT}/vendor
```
Do we need the internal clientset? It was not vendored.

DirectXMan12 replied:
it's in the original copy of `k8s.io/metrics` -- it's just not used here.  I can drop it if you want.

caesarxuchao replied:
Maybe we can even drop the internal version of the metrics API as I commented in another issue?

DirectXMan12 committed:
Fix HPA unit tests with new fake client

The new fake client properly represents the resource of `PodMetrics` as
"pods" and the resource of `NodeMetrics` as "nodes".  Previously, it
used "podmetricses" and "nodemetrics", respectively.

This fixes up `horizontal_test.go` and `replica_calc_test.go` to use the
new names.

caesarxuchao commented:
Squash and lgtm. It will conflict with https://github.com/kubernetes/kubernetes/pull/45159 though, we'll need to re-run hack/update-codegen.sh.

DirectXMan12 commented:
@caesarxuchao I've regenerated the fake client after I rebased.  Should be all good to go.  Which commits did you want squashed?  These mostly seem like distinct commits to me...

caesarxuchao commented:
@k8s-bot unit test this

caesarxuchao commented:
@k8s-bot bazel test this

caesarxuchao commented:
Regarding squashing, i'm usually requested to squash my PRs to one or two commits by my reviewers so i'm under the impression that that's the rule. I just checked the development [doc](https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#squash-and-merge), it says "all commits left on the review branch should represent meaningful milestones or units of work", so we are good to go.

caesarxuchao commented on staging/src/k8s.io/metrics/pkg/client/clientset_generated/clientset/fake/BUILD:
```diff
@@ -12,15 +12,19 @@ go_library(
         "//vendor/k8s.io/apimachinery/pkg/watch:go_default_library",
         "//vendor/k8s.io/client-go/discovery:go_default_library",
         "//vendor/k8s.io/client-go/discovery/fake:go_default_library",
-        "//vendor/k8s.io/client-go/pkg/api:go_default_library",
```
Glad to see this dependency is gone.

caesarxuchao commented:
/lgtm

caesarxuchao commented:
@jbeda could you help approve the changes? Thanks.

caesarxuchao commented:
@k8s-bot gce etcd3 e2e test this

lavalamp commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -197,7 +225,7 @@ type Fake$.type|publicPlural$ struct {
 
 var resource = `
-var $.type|allLowercasePlural$Resource = $.GroupVersionResource|raw${Group: "$.groupName$", Version: "$.version$", Resource: "$.type|allLowercasePlural$"}
+var $.type|allLowercasePlural$Resource = $.GroupVersionResource|raw${Group: "$.groupName$", Version: "$.version$", Resource: "$.resource$"}
```
Passing the type to a function (namer), as was done previously, is the correct way to use this. Please add a new namer that provides your new logic. Sorry to be a pain about this but last time this got into these generators wrong it took me ages to get it corrected.

DirectXMan12 replied:
whoops, missed that namers could get access to the tag data.  I'll rework.

DirectXMan12 committed:
Update generated clients for k8s.io/metrics

This commit updates the generated clients for k8s.io/metrics based on
the generators updated in the previous commits.

DirectXMan12 commented:
@lavalamp this look better now?  Does the new Namer need to go into gengo first or should it just live in client-gen? should it be eventually submitted to gengo?

lavalamp commented:
It should live in client-gen unless folks think it'll be used by lots of other generators.

lavalamp commented on cmd/libs/go2idl/client-gen/generators/client_generator.go:
```diff
@@ -279,3 +281,27 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
+
+// NewTagOverrideNamer creates a namer.Namer which uses the contents of the given tag as
+// the name, or falls back to another Namer if the tag is not present.
+func NewTagOverrideNamer(tagName string, fallback namer.Namer) namer.Namer {
```
Nifty.

lavalamp commented on cmd/libs/go2idl/client-gen/generators/client_generator.go:
```diff
@@ -279,3 +281,27 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
+		return nameOverride
+	}
+
+	return n.fallback.Name(t)
```
slightly concerned about silent failures, should I be?

caesarxuchao replied:
Not all types have an overridden name, so i think it's not a failure.

lavalamp commented on staging/src/k8s.io/metrics/pkg/client/clientset_generated/clientset/typed/metrics/v1alpha1/podmetrics.go:
```diff
@@ -80,8 +79,8 @@ func (c *podMetricses) List(opts v1.ListOptions) (result *v1alpha1.PodMetricsLis
 func (c *podMetricses) Watch(opts v1.ListOptions) (watch.Interface, error) {
+	opts.Watch = true
 	return c.client.Get().
-		Prefix("watch").
```
This change breaks compatibility with sufficiently old versions of kubernetes, I think? @deads2k fyi

I think this is probably OK, but can you check what version we added watch as an option. And this will need a release note and a client release note.

caesarxuchao replied:
> a release note and a client release note.

The changes was done in https://github.com/kubernetes/kubernetes/pull/41887 actually. I'll update the client-go's CHANGELOG.

This PR just lets metrics client use the latest client-gen.

deads2k replied:
> This change breaks compatibility with sufficiently old versions of kubernetes, I think? @deads2k fyi

Thanks for the heads up.  I think we supported the query param as far back as kube 1.0: https://github.com/kubernetes/kubernetes/blob/release-1.0/pkg/apiserver/resthandler.go#L251

caesarxuchao commented:
@lavalamp could you approve the change?

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/client_generator.go:
```diff
@@ -39,13 +39,15 @@ func NameSystems() namer.NameSystems {
 		"privatePlural":      namer.NewPrivatePluralNamer(pluralExceptions),
-		"allLowercasePlural": namer.NewAllLowercasePluralNamer(pluralExceptions),
+		"allLowercasePlural": lowercaseNamer,
+		"resource":           NewTagOverrideNamer("resourceName", lowercaseNamer),
```
Why there is no "resourceName" tag anywhere in this PR?

DirectXMan12 replied:
What do you mean?  `k8s.io/metrics` already makes use of the `resourceName` tag...

caesarxuchao replied:
I see. I forgot, sorry.

lavalamp commented:
/approve

caesarxuchao commented:
/lgtm

caesarxuchao commented:
@k8s-bot pull-kubernetes-node-e2e test this

caesarxuchao commented:
@DirectXMan12 @Random-Liu could you take a look at the node test failure?

```
...
I0525 13:57:15.661] SUCCESS! -- 171 Passed | 0 Failed | 0 Flaked | 0 Pending | 51 Skipped 
I0525 13:57:15.662] 
I0525 13:57:15.662] Ginkgo ran 1 suite in 20m20.179174345s
I0525 13:57:15.662] Test Suite Passed
I0525 13:57:15.662] 
I0525 13:57:15.662] Success Finished Test Suite on Host tmp-node-e2e-5a33e3fd-e2e-node-ubuntu-trusty-docker10-v2-image
I0525 13:57:15.662] <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
I0525 13:57:15.662] <                              FINISH TEST                               <
I0525 13:57:15.662] <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
I0525 13:57:15.662] 
I0525 13:57:15.663] Failure: 1 errors encountered.
W0525 13:57:15.763] exit status 1
W0525 13:57:16.417] Traceback (most recent call last):
W0525 13:57:16.418]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 130, in <module>
W0525 13:57:16.418]     ARGS.service_account,
W0525 13:57:16.418]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 94, in main
W0525 13:57:16.418]     img,
W0525 13:57:16.418]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 45, in check
W0525 13:57:16.418]     subprocess.check_call(cmd)
W0525 13:57:16.418]   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
W0525 13:57:16.418]     raise CalledProcessError(retcode, cmd)
W0525 13:57:16.419] subprocess.CalledProcessError: Command '('docker', 'run', '--rm=true', '-v', '/etc/localtime:/etc/localtime:ro', '-v', '/var/run/docker.sock:/var/run/docker.sock', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/go/src/k8s.io/kubernetes:/go/src/k8s.io/kubernetes', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/go/src/k8s.io/kubernetes/_artifacts:/workspace/_artifacts', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e@tmp/secretFiles/b65285b3-db21-4b0d-abc8-d61b621c6d9b/kubernetes-jenkins-pull-c67f9519f993.json:/service-account.json:ro', '-v', '/home/jenkins/.ssh/google_compute_engine:/root/.ssh/google_compute_engine:ro', '-v', '/home/jenkins/.ssh/google_compute_engine.pub:/root/.ssh/google_compute_engine.pub:ro', '-e', 'GCE_USER=jenkins', '-e', 'GOOGLE_APPLICATION_CREDENTIALS=/service-account.json', '-e', 'JENKINS_GCE_SSH_PRIVATE_KEY_FILE=/root/.ssh/google_compute_engine', '-e', 'JENKINS_GCE_SSH_PUBLIC_KEY_FILE=/root/.ssh/google_compute_engine.pub', '-e', 'NODE_TEST_PROPERTIES=test/e2e_node/jenkins/jenkins-pull.properties', '-e', 'NODE_TEST_SCRIPT=./test/e2e_node/jenkins/e2e-node-jenkins.sh', '-e', 'REPO_DIR=/var/lib/jenkins/workspace/pull-kubernetes-node-e2e/go/src/k8s.io/kubernetes', 'gcr.io/k8s-testimages/kubekins-node:1.7-latest')' returned non-zero exit status 1
E0525 13:57:16.426] Build failed
```

caesarxuchao commented:
Confirmed it's a test-infra issue. It's being taken care of.

caesarxuchao commented:
Eh... still not merged

caesarxuchao commented:
@k8s-bot pull-kubernetes-node-e2e test this

fejta commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this
ref: kubernetes/test-infra#2932

caesarxuchao commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

caesarxuchao commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

shashidharatd commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
